### PR TITLE
fix: resolve repo root via git worktree top across session-init, codeql-scan, github skills

### LINKED ISSUE #2344: bug(github-skill): gh act workflow test fails from linked pr-autofix worktrees
## Summary

`gh act` workflow execution can fail from pr-autofix linked worktrees because git metadata is not preserved in the shape the local workflow runner expects. Agents then need to set `SKIP_WORKFLOW_LOCAL_TEST=true` even when the workflow itself is valid.

Observed during PR #2332 autofix.

## Evidence

Agent report for PR #2332:

- `gh act` full run failed from a linked worktree due git metadata loss.
- The agent used the allowed `SKIP_WORKFLOW_LOCAL_TEST=true` bypass.
- Other validations passed and the PR merged: ADR uniqueness, targeted markdownlint, actionlint, 47 pytest tests, completion gate.

## Impact

- Local workflow validation is less useful in the exact isolated worktree layout used by pr-autofix.
- Agents can mistake an environment/tooling failure for a workflow failure.
- The bypass becomes necessary for valid workflow changes, reducing confidence.

## Suggested fix

- Reproduce `test_workflow_locally.py` / `gh act` from a linked worktree under `$HOME/.cache/pr-autofix/ai-agents.git`.
- Ensure the runner receives a usable git directory/worktree, or detect linked-worktree metadata and emit a precise environment error.
- Document when `SKIP_WORKFLOW_LOCAL_TEST=true` is allowed.

## Acceptance criteria

- Local workflow testing works from a linked pr-autofix worktree, or fails with a specific non-workflow error explaining the unsupported layout.
- The failure no longer looks like a workflow validation failure.
- Tests cover linked-worktree git metadata detection.

### LINKED ISSUE #2373: bug(codeql-scan): repo root resolution points to main checkout from linked worktrees
## Summary

`codeql-scan` resolves the main checkout root instead of the linked worktree root when run from a worktree. This makes it look for `.codeql` scripts in the wrong checkout.

Observed during PR #2354 autofix.

## Evidence

Worktree/cwd:

```text
/home/richard/src/GitHub/rjmurillo/ai-agents2/.worktrees/pr-2354
```

Command:

```bash
cd .worktrees/pr-2354
python3 .claude/skills/codeql-scan/scripts/invoke_codeql_scan.py --operation validate
```

Actual output:

```text
[x] Configuration script not found: /home/richard/src/GitHub/rjmurillo/ai-agents2/.codeql/scripts/Test-CodeQLConfig.ps1
```

Expected worktree-rooted path:

```text
/home/richard/src/GitHub/rjmurillo/ai-agents2/.worktrees/pr-2354/.codeql/scripts/Test-CodeQLConfig.ps1
```

Script involved:

```text
.claude/skills/codeql-scan/scripts/invoke_codeql_scan.py
```

Likely cause: `_get_repo_root()` uses `git rev-parse --git-common-dir`, then returns `git_common.parent`. In a linked worktree, `--git-common-dir` points at the main checkout's common git directory, not the current worktree root.

## Impact

- CodeQL validation fails from pr-autofix linked worktrees.
- Agents may skip or misdiagnose CodeQL checks for PRs using isolated worktrees.
- The tool violates the worktree isolation model used by autofix.

## Suggested fix

- Resolve repo root with `git rev-parse --show-toplevel` for the current worktree, not `--git-common-dir`.
- Add a linked-worktree regression test.
- Keep bare repo handling separate if needed.

## Acceptance criteria

- Running `invoke_codeql_scan.py --operation validate` from a linked worktree looks for `.codeql` under that worktree.
- Running from the main checkout still works.
- Test covers linked worktree root resolution.

### LINKED ISSUE #2375: Process bug: session-init misdetects repo root in linked worktrees
## Summary

`session-init` writes session logs outside the checked-out worktree when the repo is a linked worktree created from a bare repository. It derives the repo root from `git rev-parse --git-common-dir`, so `$HOME/.cache/pr-autofix/ai-agents-pr-2353` produced a log under `$HOME/.cache/pr-autofix/.agents/` instead of `$HOME/.cache/pr-autofix/ai-agents-pr-2353/.agents/`.

## Repro context

- Task: PR #2353 merge resolution in isolated worktree
- Bare repo: `$HOME/.cache/pr-autofix/ai-agents.git`
- Worktree: `$HOME/.cache/pr-autofix/ai-agents-pr-2353`
- Branch: `feat/agent-catalog-and-checkpoint`

## Command and output

```bash
cd "$HOME/.cache/pr-autofix/ai-agents-pr-2353"
python3 .claude/skills/session-init/scripts/new_session_log.py --session-number 2335 --objective "Resolve PR 2353 merge conflicts and finish PR"
```

```text
Created: /home/richard/.cache/pr-autofix/.agents/sessions/2026-06-04-session-2335-resolve-2353-merge-conflicts-finish.json
Session: 2335
Branch: feat/agent-catalog-and-checkpoint
Commit: 3bdb2aac0d
CRITICAL: Validation script not found at: /home/richard/.cache/pr-autofix/scripts/validate_session_json.py

Session log created but validation FAILED.
  File: /home/richard/.cache/pr-autofix/.agents/sessions/2026-06-04-session-2335-resolve-2353-merge-conflicts-finish.json
Fix issues and re-validate:
  python3 scripts/validate_session_json.py "/home/richard/.cache/pr-autofix/.agents/sessions/2026-06-04-session-2335-resolve-2353-merge-conflicts-finish.json"
```

Git showed the actual worktree root was different:

```text
git rev-parse --show-toplevel -> /home/richard/.cache/pr-autofix/ai-agents-pr-2353
git rev-parse --git-common-dir -> /home/richard/.cache/pr-autofix/ai-agents.git
```

## Expected behavior

Session-init should use `git rev-parse --show-toplevel` for the working tree root. `--git-common-dir` points at shared Git storage for linked worktrees and is not the repository checkout root. In isolated worktree workflows this misplaces the session log and makes validation look for scripts outside the worktree.

### LINKED ISSUE #2377: bug(github-skill): test_workflow_locally.py resolves workflow path under .claude in worktrees
## Summary

`test_workflow_locally.py` can resolve a workflow path under `.claude/.github/...` instead of the repository root `.github/...` when run from a worktree. The raw `gh act` command with the correct absolute workflow path succeeds.

Observed during PR #2365 autofix.

## Evidence

Cwd/worktree:

```text
/home/richard/src/GitHub/rjmurillo/ai-agents-pr-2365
```

Command:

```bash
python3 .claude/skills/github/scripts/test_workflow_locally.py \
  --workflow .github/workflows/nightly-cli-smoke.yml \
  --event workflow_dispatch \
  --dry-run
```

Output:

```text
Error: stat /home/richard/src/GitHub/rjmurillo/ai-agents-pr-2365/.claude/.github/workflows/nightly-cli-smoke.yml: no such file or directory
```

Expected path:

```text
/home/richard/src/GitHub/rjmurillo/ai-agents-pr-2365/.github/workflows/nightly-cli-smoke.yml
```

Actual path:

```text
/home/richard/src/GitHub/rjmurillo/ai-agents-pr-2365/.claude/.github/workflows/nightly-cli-smoke.yml
```

Control command succeeded:

```bash
gh act workflow_dispatch -W /home/richard/src/GitHub/rjmurillo/ai-agents-pr-2365/.github/workflows/nightly-cli-smoke.yml -n
# gh act dry-run ok
```

## Impact

- Local workflow validation fails even when the workflow exists and `gh act` can dry-run it.
- Agents may bypass workflow validation unnecessarily.

## Suggested fix

- Resolve `--workflow` relative to repository root, not the script or `.claude` directory.
- Add a linked-worktree test for `.github/workflows/*.yml` path resolution.

## Acceptance criteria

- `test_workflow_locally.py --workflow .github/workflows/nightly-cli-smoke.yml` resolves under repo root.
- The command works from linked worktrees.
- A regression test covers the bad `.claude/.github/...` path shape.

### LINKED ISSUE #2387: new_pr.py session-log validation reads the working tree, not the branch ref
## Problem

`new_pr.py` lists changed files via `git diff base...head` (ref-based) but then validates the session log by reading `repo_root/<session_log>` from the WORKING TREE. When the head branch is not checked out in the main checkout (worktree workflows, or creating a PR for a pushed-but-not-checked-out branch), the session log file is absent from the working tree and validation fails with the opaque "Session End validation failed", even though the log exists in the branch.

## Evidence

`.claude/skills/github/scripts/pr/new_pr.py` run_validations Validation 1: `session_log_path = os.path.join(repo_root, session_log)` then `validate_session_json.py session_log_path`. repo_root resolves to the main checkout (see the worktree note in the team memory). Workaround this session: check out each branch before calling new_pr.py.

## Suggested fix

Read the session log from the branch ref (`git show <head>:<path>`) instead of the working tree, or document the checkout requirement.

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

Five linked-worktree defects shared one root cause: tools resolved repo root via `git rev-parse --git-common-dir`, which in a LINKED worktree points at the MAIN checkout's shared `.git`. In a linked worktree `.git` is a FILE (gitdir pointer), not a directory.

The fix routes every consumer through the semantics of the existing canonical helper `scripts/github_core/repo.py:get_repo_root`, which uses `git rev-parse --show-toplevel` (correct in every layout). No new helper was created; the canonical one already existed and was correct.

## Changes

- session-init (#2375): repo-root resolvers switched to `--show-toplevel`.
- codeql-scan (#2373): repo-root resolvers switched to `--show-toplevel`; bare-repo failure still maps to exit 3.
- test_workflow_locally / run_workflow_local_test (#2377, #2344): switched from a 4-parent path count to `--show-toplevel`; parse the gitdir pointer and pass `GIT_DIR` to the `gh act` subprocess.
- new_pr (#2387): `get_repo_root` switched to `--show-toplevel`; session-log validation reads the log from `head:<path>` via `git show` into a temp file, falls back to the working-tree copy, and skips with a clear warning when neither source has it.
- Bumped project-toolkit plugin to 0.5.124 and regenerated the Copilot mirrors.

## Tests

161 source-side tests plus 29 mirror tests pass, including positive, linked-worktree, and edge cases. The regression tests assert a resolver returns the toplevel, so reverting to `--git-common-dir` fails them.

Fixes #2375
Fixes #2373
Fixes #2344
Fixes #2377
Fixes #2387

### CODE REVIEW

**Comment by rjmurillo** [comment]
-->

Disposition: umbrella PR remains the preferred carrier for codeql-scan, session-init, and github-skill root fixes, with #2412 owning the validation-script fail-closed behavior. I rebased and fixed review findings in an isolated worktree, and validated 136 targeted tests plus explicit linked-worktree exit-3 checks. Push is blocked by pre-existing local pre-push process bugs: agent drift detection fails with no agent files in this PR, and the hook reports ruff unavailable outside uv.

File: .claude/skills/github/scripts/pr/new_pr.py [hunk 1]
```diff
@@ -136,6 +146,59 @@ def _extract_validatable_session_logs(
...
+    fails but the file is present locally; yield None when neither source has
+    the content, so the caller skips validation rather than fail spuriously.
+    """
+    show = subprocess.run(
+        ["git", "show", f"{head}:{session_log}"],
+        capture_output=True,
+        text=True,
+        timeout=30,
+        env=_git_env(),
+    )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/pr/new_pr.py:175]: Fixed in 4948e67d. `git show` now decodes as UTF-8 with replacement for invalid bytes, so Windows locale defaults cannot corrupt or fail session-log validation.

File: src/copilot-cli/skills/github/scripts/pr/new_pr.py [hunk 2]
```diff
@@ -136,6 +146,59 @@ def _extract_validatable_session_logs(
...
+    fails but the file is present locally; yield None when neither source has
+    the content, so the caller skips validation rather than fail spuriously.
+    """
+    show = subprocess.run(
+        ["git", "show", f"{head}:{session_log}"],
+        capture_output=True,
+        text=True,
+        timeout=30,
+        env=_git_env(),
+    )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/pr/new_pr.py:175]: Fixed in 4948e67d in the Copilot mirror. `git show` now uses UTF-8 decoding with replacement for invalid bytes.

File: .claude/skills/github/scripts/pr/new_pr.py [hunk 3]
```diff
@@ -136,6 +146,59 @@ def _extract_validatable_session_logs(
...
+            prefix="session-log-",
+            delete=False,
+        )
+        try:
+            tmp.write(show.stdout)
+            tmp.close()
+            yield tmp.name
+        finally:
+            with contextlib.suppress(OSError):
+                os.unlink(tmp.name)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/pr/new_pr.py]: Fixed in 4948e67d. The session-log temp file now uses a context manager and the cleanup path only unlinks after the name is created, so write failures do not leak an open handle.

File: src/copilot-cli/skills/github/scripts/pr/new_pr.py [hunk 4]
```diff
@@ -136,6 +146,59 @@ def _extract_validatable_session_logs(
...
+            prefix="session-log-",
+            delete=False,
+        )
+        try:
+            tmp.write(show.stdout)
+            tmp.close()
+            yield tmp.name
+        finally:
+            with contextlib.suppress(OSError):
+                os.unlink(tmp.name)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/pr/new_pr.py]: Fixed in 4948e67d in the Copilot mirror. The temp file is written with a context manager and cleaned up safely.

File: .claude/skills/github/scripts/test_workflow_locally.py [hunk 5]
```diff
@@ -35,10 +35,71 @@ def _check_command_exists(command: str) -> str | None:
...
+    return str(gitdir)
+
+
+def _act_env(repo_root: str) -> dict[str, str]:
+    """Build the subprocess env for act, GIT_DIR-aware for linked worktrees."""
+    env = dict(os.environ)
+    gitdir = _read_worktree_gitdir(repo_root)
+    if gitdir is not None:
+        env["GIT_DIR"] = gitdir
+    return env
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/test_workflow_locally.py:112]: Fixed in 3c5df15a. `_act_env` now strips inherited `GIT_DIR`, `GIT_WORK_TREE`, `GIT_COMMON_DIR`, and `GIT_INDEX_FILE`, then sets the linked-worktree `GIT_DIR` when needed.

File: src/copilot-cli/skills/github/scripts/test_workflow_locally.py [hunk 6]
```diff
@@ -35,10 +35,71 @@ def _check_command_exists(command: str) -> str | None:
...
+    return str(gitdir)
+
+
+def _act_env(repo_root: str) -> dict[str, str]:
+    """Build the subprocess env for act, GIT_DIR-aware for linked worktrees."""
+    env = dict(os.environ)
+    gitdir = _read_worktree_gitdir(repo_root)
+    if gitdir is not None:
+        env["GIT_DIR"] = gitdir
+    return env
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/test_workflow_locally.py:112]: Fixed in 3c5df15a in the Copilot mirror. `_act_env` strips hook git variables before injecting linked-worktree `GIT_DIR`.

File: scripts/validation/run_workflow_local_test.py [hunk 7]
```diff
@@ -133,6 +140,45 @@ def _run(cmd: list[str], *, timeout: int, cwd: Path | None = None) -> tuple[int,
...
+    return str(gitdir)
+
+
+def _act_env(repo_root: Path) -> dict[str, str]:
+    """Build the subprocess env for gh act, GIT_DIR-aware for linked worktrees."""
+    env = dict(os.environ)
+    gitdir = _read_worktree_gitdir(repo_root)
+    if gitdir is not None:
+        env["GIT_DIR"] = gitdir
+    return env
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/run_workflow_local_test.py:240]: Fixed in 3c5df15a. The validation runner now scrubs inherited git hook variables before running `gh act`, while preserving valid linked-worktree `GIT_DIR` injection.

File: src/copilot-cli/skills/github/scripts/test_workflow_locally.py [hunk 8]
```diff
@@ -35,10 +35,71 @@ def _check_command_exists(command: str) -> str | None:
...
+    Falls back to the four-parents path only when git is unavailable, so the
+    workflow-path resolution below still has a best-effort anchor.
+    """
+    result = subprocess.run(
+        ["git", "rev-parse", "--show-toplevel"],
+        capture_output=True,
+        text=True,
+        timeout=10,
+        check=False,
+    )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/test_workflow_locally.py]: Fixed in 3c5df15a in the Copilot mirror. `_get_repo_root` now calls git with the scrubbed environment so hook overrides cannot redirect root resolution.

File: .claude/skills/github/scripts/test_workflow_locally.py [hunk 9]
```diff
@@ -35,10 +35,71 @@ def _check_command_exists(command: str) -> str | None:
...
+    Falls back to the four-parents path only when git is unavailable, so the
+    workflow-path resolution below still has a best-effort anchor.
+    """
+    result = subprocess.run(
+        ["git", "rev-parse", "--show-toplevel"],
+        capture_output=True,
+        text=True,
+        timeout=10,
+        check=False,
+    )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/test_workflow_locally.py]: Fixed in 3c5df15a. `_get_repo_root` now strips git hook override variables before calling `git rev-parse --show-toplevel`.

File: src/copilot-cli/skills/github/scripts/pr/new_pr.py [hunk 10]
```diff
@@ -136,6 +146,59 @@ def _extract_validatable_session_logs(
...
+    if os.path.exists(worktree_path):
+        yield worktree_path
+        return
+
+    print(
+        f"  WARNING: session log {session_log} not found at {head} or in the "
+        "working tree; skipping Session End validation.",
+        file=sys.stderr,
+    )
+    yield None
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/pr/new_pr.py]: Fixed in 4948e67d in the Copilot mirror. `_session_log_for_validation` no longer validates a working-tree fallback when `head:<path>` is missing, so stale local files cannot produce a false pass.

File: src/copilot-cli/skills/github/scripts/test_workflow_locally.py [hunk 11]
```diff
@@ -35,10 +35,71 @@ def _check_command_exists(command: str) -> str | None:
...
+    Uses git rev-parse --show-toplevel rather than counting parent
+    directories from __file__. Path-counting is fragile to script relocation
+    (it resolved under .claude/, not the repo root) and wrong in a LINKED
+    worktree, where the script may be vendored at a different depth (#2377).
+    --show-toplevel returns the current worktree root in every layout.
+    Canonical reference: scripts/github_core/repo.py::get_repo_root.
+
+    Falls back to the four-parents path only when git is unavailable, so the
+    workflow-path resolution below still has a best-effort anchor.
+    """
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/test_workflow_locally.py]: Fixed in 3c5df15a in the Copilot mirror. The docstring now states the fallback applies when git cannot return a worktree root, matching the implementation.

File: src/copilot-cli/.claude-plugin/plugin.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "project-toolkit",
   "description": "Specialized agent definitions, reusable skills, and lifecycle hooks for GitHub Copilot CLI, generated from Claude canonical sources",
-  "version": "0.5.120",
+  "version": "0.5.121",
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/.claude-plugin/plugin.json]: Fixed by updating the PR description after the merge-main bump. The body now states project-toolkit 0.5.121, matching the manifest.

File: .claude/skills/github/scripts/pr/new_pr.py [hunk 13]
```diff
@@ -136,6 +146,59 @@ def _extract_validatable_session_logs(
...
+        os.makedirs(scratch_dir, exist_ok=True)
+        tmp_name = ""
+        try:
+            with tempfile.NamedTemporaryFile(
+                mode="w",
+                encoding="utf-8",
+                suffix=".json",
+                prefix=".session-log-",
+                dir=scratch_dir,
+                delete=False,
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/pr/new_pr.py:189]: Fixed in da1ce692. Session-log validation scratch now writes under ignored `.agents/scratch/session-log-validation/`, not tracked `.agents`, and cleanup still unlinks the file after validation. `tests/test_new_pr.py` passed.

File: .claude/skills/github/scripts/pr/new_pr.py [hunk 14]
```diff
@@ -63,23 +66,30 @@ def _git_env() -> dict[str, str]:
 
 
 def get_repo_root() -> str:
-    """Get the git repository root directory."""
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/pr/new_pr.py:66]: Fixed in ee8f1a25. `_git_env` now strips `GIT_DIR`, `GIT_WORK_TREE`, `GIT_COMMON_DIR`, and `GIT_INDEX_FILE`, matching the workflow helpers. `tests/test_new_pr.py` passed.

File: .claude/skills/github/scripts/test_workflow_locally.py [hunk 15]
```diff
@@ -35,10 +35,80 @@ def _check_command_exists(command: str) -> str | None:
...
+    return str(gitdir)
+
+
+def _act_env(repo_root: str) -> dict[str, str]:
+    """Build the subprocess env for act, GIT_DIR-aware for linked worktrees."""
+    env = _git_env()
+    gitdir = _read_worktree_gitdir(repo_root)
+    if gitdir is not None:
+        env["GIT_DIR"] = gitdir
+    return env
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/test_workflow_locally.py:112]: Fixed in 84d878f6 in the canonical. The workflow-local runner now checks linked-worktree gitdir pointers before invoking act and returns a clear exit-2 setup error when the gitdir is missing. `tests/test_test_workflow_locally.py` passed.

File: src/copilot-cli/skills/github/scripts/test_workflow_locally.py [hunk 16]
```diff
@@ -35,10 +35,80 @@ def _check_command_exists(command: str) -> str | None:
...
+    return str(gitdir)
+
+
+def _act_env(repo_root: str) -> dict[str, str]:
+    """Build the subprocess env for act, GIT_DIR-aware for linked worktrees."""
+    env = _git_env()
+    gitdir = _read_worktree_gitdir(repo_root)
+    if gitdir is not None:
+        env["GIT_DIR"] = gitdir
+    return env
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply src/copilot-cli/skills/github/scripts/test_workflow_locally.py:112]: Fixed in 84d878f6 in the Copilot mirror. The workflow-local runner now checks linked-worktree gitdir pointers before invoking act and returns a clear exit-2 setup error when the gitdir is missing. `tests/test_test_workflow_locally.py` passed.

File: .claude/skills/github/scripts/test_workflow_locally.py [hunk 17]
```diff
@@ -35,10 +35,92 @@ def _check_command_exists(command: str) -> str | None:
...
+    result = subprocess.run(
+        ["git", "rev-parse", "--show-toplevel"],
+        capture_output=True,
+        text=True,
+        timeout=10,
+        check=False,
+        env=_git_env(),
+    )
+    if result.returncode == 0 and result.stdout.strip():
+        return str(Path(result.stdout.strip()).resolve())
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/github/scripts/test_workflow_locally.py:62]: Fixed in 595a9c85. `_get_repo_root` now catches `FileNotFoundError` and `subprocess.TimeoutExpired` and uses the documented four-parent fallback. `tests/test_test_workflow_locally.py` passed.
